contracts/mixnet/src/transactions.rs

```diff
@@ -693,9 +725,32 @@ pub(crate) fn try_remove_delegation_from_mixnode(
     deps: DepsMut,
     info: MessageInfo,
     mix_identity: IdentityKey,
+) -> Result<Response, ContractError> {
+    let sender = info.sender.clone();
+    _try_remove_delegation_from_mixnode(deps, info, mix_identity, &sender)
+}
+
+pub(crate) fn try_remove_delegation_from_mixnode_on_behalf(
+    deps: DepsMut,
+    info: MessageInfo,
+    mix_identity: IdentityKey,
+    delegate_addr: Addr,
+) -> Result<Response, ContractError> {
+    if info.sender != VESTING_CONTRACT_ADDR {
+        return Err(ContractError::Unauthorized);
+    }
+
+    _try_remove_delegation_from_mixnode(deps, info, mix_identity, &delegate_addr)
+}
+
+fn _try_remove_delegation_from_mixnode(
```

Comment:
I've been thinking through that, undelegations also have to go back to the vesting contract, as ppl can delegate unvested coins, so we can't let them withdraw them. So the flow is for folks to execute against the vesting contract and then withdraw what they can once the funds get back into it.

Reply:
But, unless I misunderstood it, the delegation information [inside the mixnet contract] is saved using the "delegator's" address. So in principle nothing is preventing the following to happen:
- user delegates bunch of tokens to some mixnode using the vesting contract
- now he calls the mixnet contract directly to undelegate it all
- he skipped all the vesting

Reply:
Thats a beautiful workaround, yeah, I think the best is to store it in a different bucket, love that paranoid mind of yours :)

Reply:
@jstuczyn Thinking it through some more, saving them to different buckets would introduce complexity throughout the stack, as we'd have to double everything we're doing for rewarding as well, I'm thinking of putting in a flag into the RawDelegationData that specifies if the delegation can be withdrawn by the owner...

Reply:
Thinking it through some more, that would entail a migration, and a lot of updates, still not sure which path to take

Reply:
Since an additional problem of the undelegation procedure is that we have to know where to send the funds, and in the case of `proxy` delegations this is the vesting account, I think it makes most sense to add an `Option<Addr>` field to the `RawDelegationsData` struct. That way we can both know if the delegation is a proxy delegation, and where to send the funds on undelegate...

Reply:
@jstuczyn I'm pretty happy with how this looks now, let me know what you think